### PR TITLE
THIRDEYE-856 : Create monitor module in distributed anomaly detection

### PR DESCRIPTION
**Pull request by npawar**
This pull request contains the monitor module. The monitor module periodically(period configurable, default 1 HOUR) reads anomaly jobs and anomaly tasks. When it finds anomaly jobs which have all of their tasks in COMPLETED state, it updates the anomaly job to status COMPLETED. This module also checks for records in both tables, which are COMPLETED and not updated for over a configurable number of days(default 7 DAY), and deletes them.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.linkedin.thirdeye.anomaly.monitor;
+
+public class MonitorConfiguration {
+
+  private int expireDaysAgo = 7;
```
Comment (raviarin) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorConfiguration.java]: Make 7 and 1 as constants / default values to MonitorConstants class

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  @Override
+  public void run() {
+    try {
+      LOG.info("Starting monitor job");
+
+      List<AnomalyJobSpec> anomalyJobSpecs = findAnomalyJobsWithStatusScheduled();
+      jobContext.setJobName(TaskType.MONITOR.toString());
+      long jobExecutionId = createAnomalyJob();
```
Comment (raviarin) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java]: Rename CreateAnomalyJob to CreateMonitorJob

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+      jobContext.setJobExecutionId(jobExecutionId);
+      List<Long> taskIds = createAnomalyTasks(anomalyJobSpecs);
```
Comment (raviarin) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java]: Rename to CreateMonitorTaskId

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+  private long createAnomalyJob() {
+    Session session = sessionFactory.openSession();
+    Long jobExecutionId = null;
+    try {
+      ManagedSessionContext.bind(session);
+      Transaction transaction = session.beginTransaction();
+      try {
+        LOG.info("Creating monitor job");
+        AnomalyJobSpec anomalyJobSpec = new AnomalyJobSpec();
```
Comment (raviarin) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java]: Trying having a parent class called JobSpec. AnomalyJobSpec and MonitorJobSpec should inherit from the parent class

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyConfiguration.java [hunk 5]
```diff
@@ -1,14 +1,17 @@
 package com.linkedin.thirdeye.anomaly;
 
+import com.linkedin.thirdeye.anomaly.monitor.MonitorConfiguration;
 import com.linkedin.thirdeye.common.ThirdEyeConfiguration;
 import com.linkedin.thirdeye.detector.driver.FailureEmailConfiguration;
 
 public class ThirdEyeAnomalyConfiguration extends ThirdEyeConfiguration {
   private boolean scheduler = false;
   private boolean worker = false;
+  private boolean monitor = false;
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyConfiguration.java]: Can an executor be more than one of (scheduler / worker / monitor) ? 
if no then extract these 3 booleans to type.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+import com.linkedin.thirdeye.anomaly.job.JobContext;
+import com.linkedin.thirdeye.anomaly.task.TaskConstants.TaskStatus;
+import com.linkedin.thirdeye.anomaly.task.TaskConstants.TaskType;
+import com.linkedin.thirdeye.anomaly.task.TaskGenerator;
+import com.linkedin.thirdeye.detector.db.dao.AnomalyJobSpecDAO;
+import com.linkedin.thirdeye.detector.db.dao.AnomalyTaskSpecDAO;
+import com.linkedin.thirdeye.detector.db.entity.AnomalyJobSpec;
+import com.linkedin.thirdeye.detector.db.entity.AnomalyTaskSpec;
+
+public class MonitorJobRunner implements Runnable {
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java]: Can we have a JobRunner interface or abstract class ?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorTaskRunner.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      }
+    } finally {
+      session.close();
+      ManagedSessionContext.unbind(sessionFactory);
+    }
+  }
+
+  private void executeMonitorExpire(MonitorTaskInfo monitorTaskInfo) {
+    LOG.info("Execute monitor expire {}", monitorTaskInfo);
+    Session session = sessionFactory.openSession();
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorTaskRunner.java]: Use @UnitOfWork and avoid creating manual transactions.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/task/TaskExecutorFactory.java [hunk 8]
```diff
@@ -8,11 +10,12 @@
     Class<?> taskExecutorClass = null;
     switch(taskType) {
       case ANOMALY_DETECTION:
-        taskExecutorClass = AnomalyDetectionTaskRunner.class;
+        taskExecutorClass = DetectionTaskRunner.class;
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/task/TaskExecutorFactory.java]: Why returning class when you can return the object itself. 
This is TaskExecutorFactory which should create TaskExecutor instances.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorConfiguration.java (hunk 1 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorConfiguration.java]: Fixed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyConfiguration.java (hunk 5 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyConfiguration.java]: Yes a node can be more than just a scheduler/worker/monitor

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java (hunk 6 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java]: Yes good suggestion, fixed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorTaskRunner.java (hunk 7 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorTaskRunner.java]: thanks for the tip!

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/task/TaskExecutorFactory.java (hunk 8 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/task/TaskExecutorFactory.java]: Fixed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionTaskRunner.java [hunk 9]
```diff
@@ -185,65 +193,57 @@ public AnomalyDetectionTaskRunner() {
...
       session.close();
       ManagedSessionContext.unbind(sessionFactory);
     }
-
     return results;
   }
 
+  @UnitOfWork
   private void handleResults(List<AnomalyResult> results) {
     Session session = sessionFactory.openSession();
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionTaskRunner.java]: No need to open session when you have unitOfWork annotation.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionTaskRunner.java [hunk 10]
```diff
@@ -185,65 +193,57 @@ public AnomalyDetectionTaskRunner() {
...
+        result.setMetric(spec.getMetric());
+        result.setFilters(spec.getFilters());
+
+        // make sure score and weight are valid numbers
+        result.setScore(normalize(result.getScore()));
+        result.setWeight(normalize(result.getWeight()));
+        resultDAO.save(result);
       }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionTaskRunner.java]: Catching an Exception which is RTE and wrapping it again in RTE and throwing it ? 
No need to catch this in the first place.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/job/JobRunner.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package com.linkedin.thirdeye.anomaly.job;
+
+import java.util.List;
+
+public interface JobRunner extends Runnable {
+
+  long createJob();
+  List<Long> createTasks();
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/job/JobRunner.java]: Please add javadoc.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+      List<Long> taskIds = createTasks();
+
+    } catch (Exception e) {
+      LOG.error("Exception in monitor job runner", e);
+    }
+  }
+
+  @UnitOfWork
+  public long createJob() {
+    Session session = sessionFactory.openSession();
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java]: Not needed.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+      LOG.info("Creating monitor job");
+      AnomalyJobSpec anomalyJobSpec = new AnomalyJobSpec();
+      anomalyJobSpec.setJobName(monitorJobContext.getJobName());
+      anomalyJobSpec.setScheduleStartTime(System.currentTimeMillis());
+      anomalyJobSpec.setStatus(JobStatus.SCHEDULED);
+      jobExecutionId = anomalyJobSpecDAO.save(anomalyJobSpec);
+      LOG.info("Created anomalyJobSpec {} with jobExecutionId {}", anomalyJobSpec,
+          jobExecutionId);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java]: catching RTE and wrapping it in RTE and throwing not needed.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+      session.close();
+      ManagedSessionContext.unbind(sessionFactory);
+    }
+    return jobExecutionId;
+  }
+
+
+  @UnitOfWork
+  public List<Long> createTasks() {
+    Session session = sessionFactory.openSession();
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/monitor/MonitorJobRunner.java]: same as above.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/task/TaskDriver.java [hunk 15]
```diff
@@ -153,25 +150,18 @@ private AnomalyTaskSpec selectAndUpdate() throws Exception {
...
       LOG.info(Thread.currentThread().getId() + " : update status {}", updateStatus);
 
-      if (!transaction.wasCommitted()) {
-        transaction.commit();
-      }
     } catch (Exception e) {
-      if (transaction != null) {
-        transaction.rollback();
-      }
+      throw new RuntimeException(e);
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/task/TaskDriver.java]: same as above.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/dao/AnomalyJobSpecDAO.java [hunk 16]
```diff
@@ -14,4 +18,33 @@ public AnomalyJobSpecDAO(SessionFactory sessionFactory) {
...
+  public boolean updateStatusAndJobEndTime(Long id, JobStatus status, Long jobEndTime) {
+    try {
+      int numRowsUpdated = namedQuery("com.linkedin.thirdeye.anomaly.AnomalyJobSpec#updateStatusAndJobEndTime")
+          .setParameter("id", id)
+          .setParameter("status", status)
+          .setParameter("jobEndTime", jobEndTime)
+          .executeUpdate();
+      return numRowsUpdated == 1;
+    } catch (HibernateException exception) {
+      exception.printStackTrace();
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/dao/AnomalyJobSpecDAO.java]: please use LOGGER.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionTaskRunner.java (hunk 9 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionTaskRunner.java]: That doesn't seem to work "org.hibernate.HibernateException: No session currently bound to execution context" 
Remove the session handling and add it somewhere else?

**Comment by npawar** [comment]
Fixed all comments except session handling.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/task/TaskDriver.java [hunk 17]
```diff
@@ -139,7 +139,7 @@ private AnomalyTaskSpec selectAndUpdate() throws Exception {
         }
         Thread.sleep(1000);
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/task/TaskDriver.java]: Why sleeping here ?

**Comment by puneetjaiswal** [comment]
LGTM

**Comment by npawar** [comment]
Testing done: Started the monitor module in conjunction with the schedulers and workers.
Tested that a monitor job is created at every scheduled run of the monitor. Checked that a monitor task is created for every job in the jobs table with status != COMPLETED. Checked that an additional monitor task of type expire is created. The monitor task checks the status of the job and the corresponding tasks and updates it accordingly. The expire task expires all jobs and tasks which are older than a predefined configurable retention

**Pull request merged**